---
title: "Differential Gene Expression Analysis"
teaching: 0
exercises: 0
questions:
- "How do I find genes differentially expressed between treatment groups?"
objectives:
- "To check for sample mixup in data."
- "To identify genes differentially expressed between treatment groups."
- "To describe the effect of sample size on differential gene expression analysis."
keypoints:
- "Non-coding RNA can be used to check for sample mixups in RNA sequencing data."
source: Rmd
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("02-")
```

One of the most common applications of RNA sequencing technology is to identify genes that are differentially expressed between sample groups, for example, between wild type and mutant, or between tumor and normal samples. Count data report the number of sequence reads (fragments) assigned to each gene, which describes the expression abundance of a gene. Similar data can be found in ChIP-Seq, HiC, shRNA screening, or mass spectrometry experiments.

![](../fig/RNAseq-workflow.png)

Once we have aligned sequence reads and have quantified expression, we can continue the pipeline with differential expression analysis. We will use read counts at the gene level and the R package [DESeq2](http://bioconductor.org/packages/release/bioc/html/DESeq2.html). 

R Libraries and Data Import
------------------------------------
Load the R libraries.

```{r message=FALSE}
library("DESeq2")
library(ggplot2)
library(dplyr)
getwd()
```

Load the data files.

```{r load_data}
# Load raw read counts.
load("../data/DO192_RNAseq_EMASE_RawCounts.Rdata")

# Load covariates & annotations.
load("../data/ChickMungeretal2016_DiversityOutbred.Rdata")

# a quick look at the expression data
dim(expr.rna.192.rawcounts)
names(expr.rna.192.rawcounts)
expr.rna.192.rawcounts[1:5, 1:5]
```

The raw read counts have `r nrow(expr.rna.192.rawcounts)` rows and `r ncol(expr.rna.192.rawcounts)` columns, with animal IDs for row names and Ensembl gene IDs for column names. The first five rows and five columns are shown above.

Create a data frame containing key experimental design factors for this experiment. These factors include diet and sex.

```{r exp_design}
exp_design <- data.frame(
  mouseIDs = rownames(expr.rna.192.rawcounts),
  diet = covariates.rna.192$Diet,
  sex = covariates.rna.192$Sex,
  coat_color = covariates.rna.192$Coat.Color)
head(exp_design)
```

Check to make sure that all mouse IDs are represented in the raw read counts and in the experimental design file.  

```{r check_data}
all(rownames(expr.rna.192.rawcounts)==exp_design$mouseIDs)
```

A quick check for sample mixup
------------------------------
Let's do a quick check for sample mixup with *Xist* gene expression. *Xist*, or X-inactive specific transcript, produces non-protein coding RNA. The gene is expressed exclusively from the inactivated X chromosome in female mice.  

```{r check_xist}
geneID <- "ENSMUSG00000086503"
geneName <- "Xist"
gIndex <- which(colnames(expr.rna.192.rawcounts)==geneID)
data <- data.frame(exp_design,
                   exp=as.numeric(expr.rna.192.rawcounts[,gIndex]))
```


Plot **Xist** expression in all samples against sex.   

```{r plot_xist}
p <- ggplot(data, aes(x = sex, y = exp)) 
p <- p + geom_point(position = position_jitter(width = 0.1), 
                    size = 3, aes(colour = sex))
p <- p + stat_summary(fun.y = mean, geom = "point")
p <- p + ylab(label = "Gene Expression (Read Counts)")
p <- p + ggtitle(label = "Xist", subtitle = "ENSMUSG00000086503")
p
```

Female mice averaged 15,000 raw read counts for *Xist*, while males had none. We can rest assured that male and female samples weren't mixed up.

We'll start with an example identifying the genes that are differentially expressed in the samples between two diets **Chow** and **High fat**.

```{r sample_ind}
male_index = which(exp_design$sex=="M")
female_index = which(exp_design$sex=="F")
chow_index= which(exp_design$diet=="chow")
hf_index= which(exp_design$diet=="HF")
male_chow = intersect(male_index, chow_index)
male_hf = intersect(male_index, hf_index)
```

Differential Expression Analysis with **three** samples in each group
------------------------------------------------------------------
To make the example simple, we'll subset the expression data such that we have **3 DO mice** under **Chow diet** and 3 DO mice under **High Fat diet**.  

```{r sample_size}
sampleSize = 3
```

Later on we will see the effect of sample size by varying it.  

```{r subset_exp}
diet_DE <- c(male_chow[1:sampleSize], male_hf[1:sampleSize])
exp_design_diet_DE <- exp_design[diet_DE,]
exp_design_diet_DE
exp_diet_DE <- expr.rna.192.rawcounts[diet_DE,]
all(rownames(exp_diet_DE) == as.vector(exp_design_diet_DE$mouseIDs))
exp_diet_DE[, 1:5]
```

Filter out genes with zero and low expression (fewer than 5 read counts) in 50% of the samples.

```{r filter_exp}
thres <- 5
nzIndex <- as.vector(which(apply(exp_diet_DE, 1, function(x){sum(x>thres)/length(x)})>=0.5))
head(nzIndex)
exp.dietDE = exp_diet_DE[nzIndex,]
dim(exp.dietDE)
```

Create data frames for **DESeq2** object.  

```{r dataframe_deseq2, results="hide"}
### colData contains the condition/group information for Differenetial expression analysis
colData <- DataFrame(group = factor(exp_design_diet_DE$diet))
```
```{r deseq2_obj, results="hide"}
### Create DESeq2 object using expression and colData
dds <- DESeqDataSetFromMatrix(countData = as.data.frame(round(exp.dietDE)),
         colData = colData, design = ~ group)
dds <- DESeq(dds)
res = results(dds)
```
```{r res_summary, results="hide"}
### summary of Differential Expression analysis
summary(res)
```
```{r MA_plot, results="hide"}
plotMA(res, main="M-A Plot: 3 Samples per group", ylim=c(-2,2))
```
```{r gene_plot, results="hide"}
d<-plotCounts(dds, gene=which.min(res$padj), intgroup="group",
              returnData=TRUE)
p <- ggplot(d, aes(x=group, y=count)) +
  geom_point(position=position_jitter(w=0.2,h=0),size=3)
p <- p + theme(axis.text=element_text(size=12),
               axis.title=element_text(size=20,face="bold", colour = "blue"),
               plot.title = element_text(size = rel(2)))
p
```
Let us plot the histogram of p-values. The p-value histogram is a good diagnostic test for the differential expression analysis.

```{r pval_hist, results="hide"}
hist(res$pvalue,breaks=100,col="grey", xlab="p-value",main="p-value histogram: 3 Samples per group")
```

Differential Expression Analysis with **ten** samples in each **diet** group
------------------------------------------------------------------

```{r sample_size_10, results="hide"}
sampleSize = 10
```
Later on we will see the effect of sample size by varying it.
```{r subset_exp_10, results="hide"}
diet_DE = c(male_chow[1:sampleSize],male_hf[1:sampleSize])
exp_design_diet_DE= exp_design[diet_DE,]
exp_design_diet_DE
exp_diet_DE=expr.rna.192.rawcounts[,diet_DE]
all(colnames(exp_diet_DE)==as.vector(exp_design_diet_DE$mouseIDs))
head(exp_diet_DE)
```
```{r head_exp_10}
head(exp_diet_DE)
```

Let us filter out genes with zero and low expression (less than 5 read counts) in 50% of the samples.
```{r filter_exp_10, results="hide"}
thres= 5
nzIndex= as.vector(which(apply(exp_diet_DE,1,function(x){sum(x>thres)/length(x)})>=0.5))
head(nzIndex)
exp.dietDE = exp_diet_DE[nzIndex,]
dim(exp.dietDE)
```
Let us create data frames for DESeq2 object
```{r dataframe_deseq2_10, results="hide"}
### colData contains the condition/group information for Differenetial expression analysis
colData <- DataFrame(group = factor(exp_design_diet_DE$diet))
```
```{r deseq2_obj_10, results="hide"}
### Create DESeq2 object using expression and colData
dds <- DESeqDataSetFromMatrix(countData = as.data.frame(round(exp.dietDE)),
         colData = colData, design = ~ group)
dds <- DESeq(dds)
res = results(dds)
```
```{r res_summary_10, results="hide"}
### summary of Differential Expression analysis
summary(res)
```
```{r MA_plot_10, results="hide"}
plotMA(res, main="M-A Plot: 10 Samples per group", ylim=c(-2,2))
```
```{r gene_plot_10, results="hide"}
d<-plotCounts(dds, gene=which.min(res$padj), intgroup="group",
              returnData=TRUE)
p <- ggplot(d, aes(x=group, y=count)) +
  geom_point(position=position_jitter(w=0.2,h=0),size=3)
p <- p + theme(axis.text=element_text(size=12),
               axis.title=element_text(size=20,face="bold", colour = "blue"),
               plot.title = element_text(size = rel(2)))
p
```
```{r pval_hist_10, results="hide"}
hist(res$pvalue,breaks=100,col="grey", xlab="p-value",main="p-value Histogram: 10 Samples per group")
```

```{r svd_pca, results="hide"}
svd.obj = svd(apply(exp.dietDE,1,function(x){x-mean(x)}))
plot(svd.obj$d^2/sum(svd.obj$d^2),ylab="Percent Variance Explained", main="PC of expression data")
```
```{r svd_cor, results="hide"}
print(cor(svd.obj$u[,1],as.numeric(as.factor(exp_design_diet_DE$diet))))
print(cor(svd.obj$u[,2],as.numeric(as.factor(exp_design_diet_DE$diet))))
print(cor(svd.obj$u[,5],as.numeric(as.factor(exp_design_diet_DE$diet))))
print(cor(svd.obj$u[,1],
    as.numeric(as.factor(covariates.rna.192$Coat.Color[diet_DE]))))
print(cor(as.numeric(as.factor(exp_design_diet_DE$diet)),
                         as.numeric(as.factor(covariates.rna.192$Coat.Color[diet_DE]))))
```
